docs/cli-roadmap.md

```diff
@@ -0,0 +1,82 @@
+  1. label updates: addlabel, rmlabel, changelabel
+  1. annotation updates: addannotation, rmannotation, changeannotation
+  1. more user-friendly on-command-line json for patch
+1. We also want to support the following flavors of more general updates:
+  1. whichever we don’t support:
+    1. safe update: update the full resource, guarded by resourceVersion precondition (and perhaps selected value-based preconditions)
+    1. forced update: update the full resource, blowing away the previous Spec without preconditions; delete and re-create if necessary
+  1. diff/dryrun: Compare new config with current Spec
+  1. submit/apply/reconcile/ensure/merge: Merge user-provided fields with current Spec. Keep track of user-provided fields using an annotation -- see [#1702](https://github.com/GoogleCloudPlatform/kubernetes/issues/1702). Delete all objects with deployment-specific labels.
+1. --dry-run for all commands
+1. Support full label selection syntax, including support for namespaces.
+1. Wait on conditions [#1899](https://github.com/GoogleCloudPlatform/kubernetes/issues/1899)
+1. Make kubectl scriptable: make output and exit code behavior consistent and useful for wrapping in workflows and piping back into kubectl and/or xargs (e.g., dump full URLs?, distinguish permanent and retry-able failure, identify objects that should be retried)
+1. Easy generation of clean configuration files from existing objects (including containers -- podex) -- remove readonly fields, status
+  1. Export from one namespace, import into another is an important use case
+1. Derive objects from other objects
+  1. pod clone
+  1. rc from pod
+  1. --labels-from (services from pods or rcs)
+1. Kind discovery (i.e., operate on objects of all kinds) [#3233](https://github.com/GoogleCloudPlatform/kubernetes/pull/3233)
+1. A fairly general-purpose way to specify fields on the command line during creation and update, not just from a config file
+1. Extensible API-based generator framework, so that complex client libraries don’t need to be rewritten in multiple languages, and so that the abstractions are available through all interfaces: API, CLI, UI, logs, ...
+  1. Need schema registry, and some way to invoke generator (e.g., using a container)
+  1. Convert run-container to API-based generator
+1. Transformation framework
+  1. More intelligent defaulting of fields (e.g., [#2643](https://github.com/GoogleCloudPlatform/kubernetes/issues/2643))
+1. Update preconditions based on the values of arbitrary object fields. 
+
+TODO: 
+* watch
+* ssh [#1513](https://github.com/GoogleCloudPlatform/kubernetes/issues/1513)
+* attach [#1521](https://github.com/GoogleCloudPlatform/kubernetes/issues/1521)
+* image/registry commands
+* do any other server paths make sense? validate? generic curl functionality?
+* template parameterization
+* dynamic/runtime configuration
+
+Server-side support:
+
+1. Finish v1beta3
```

Comment:
Note: Defaulting of selectors from labels: https://github.com/GoogleCloudPlatform/kubernetes/issues/1698#issuecomment-71048278